### ISSUE TITLE
Finetuning dataloader does not respect shuffle seed

### ISSUE THREAD

**Issue by janEbert** [mosaicml/llm-foundry#1676]
In `build_finetuning_dataloader`, either a [`StreamingDataset` is created based on the supplied configuration](https://github.com/mosaicml/llm-foundry/blob/bd113da06a9ca66890d628a9afc8b5c5ea7f02b1/llmfoundry/data/finetuning/dataloader.py#L255-L269) or a [`datasets.Dataset`](https://github.com/mosaicml/llm-foundry/blob/bd113da06a9ca66890d628a9afc8b5c5ea7f02b1/llmfoundry/data/finetuning/dataloader.py#L300-L312). For the `StreamingDataset`, its constructor arguments like `shuffle` and `shuffle_seed` are simply taken from the configuration with some minor filtering.
For the `datasets.Dataset`'s shuffling, afterwards, a [`DistributedSampler` is created](https://github.com/mosaicml/llm-foundry/blob/bd113da06a9ca66890d628a9afc8b5c5ea7f02b1/llmfoundry/data/finetuning/dataloader.py#L331-L339) which is passed the `shuffle` config value. However, it does not receive a potential `shuffle_seed`, meaning its shuffling is cannot be adjusted and will be deterministic across all runs.

(Below is old, incorrect text for the sake of @dakinggg's response making sense.)

~~In `build_finetuning_dataloader`, a [`StreamingDataset` is created based on the supplied configuration](https://github.com/mosaicml/llm-foundry/blob/bd113da06a9ca66890d628a9afc8b5c5ea7f02b1/llmfoundry/data/finetuning/dataloader.py#L300-L312). `StreamingDataset`'s constructor arguments like `shuffle` are simply taken from the configuration with some minor filtering. However, afterwards, a [`DistributedSampler` is created](https://github.com/mosaicml/llm-foundry/blob/bd113da06a9ca66890d628a9afc8b5c5ea7f02b1/llmfoundry/data/finetuning/dataloader.py#L334) which is again passed the `shuffle` config value. Since the `StreamingDataset` already received a value for `shuffle`, this is redundant.~~

~~Nothing is really wrong, but the second shuffling is simply not required.~~

**Comment by dakinggg** [comment]
The variable name is definitely confusing here...but the sampler is actually created only in the case that we have a huggingface dataset (its in the `else` branch) which does not get shuffled before the sampler is applied.

**Comment by janEbert** [comment]
You're absolutely right; sorry about that. I actually found the same and already implemented what I proposed in https://github.com/mosaicml/composer/issues/3723, but got distracted while looking over it and forgot... :)
I edited the issue to focus on the `seed` behavior in this case.
